### ISSUE TITLE
(discussion) Authentication Service

### ISSUE THREAD

**Issue by trainerbill** [meanjs/mean#1129]
I think we need an update to the authentication service.

https://github.com/meanjs/mean/blob/master/modules/users/client/services/authentication.client.service.js

Not sure why we would want to read out of window.user.  In one of my apps I made an authentication service with a promise:

```js
(function() {
  'use strict';

  angular
    .module('users')
    .factory('Authentication', Authentication);

  Authentication.$inject = ['$resource', '$q'];
  function Authentication($resource, $q) {

    var promise = $q.defer();

    var service = {
      user: null,
      refresh: refresh,
      signin: $resource('api/v1/auth/signin'),
      signout: $resource('api/v1/auth/signout')
    };

    function refresh() {
      service.user = $resource('api/v1/users/me')
                      .get();
      return service.user.$promise;

    }

    refresh().then(function () {
      promise.resolve(service);
    });


    return promise.promise;
  }
})();
```

Then I just hit the /users/me endpoint to get the user information.  I am thinking this is probably a better way.  It took awhile to change a lot of the application to work with it so I wanted some thoughts before I submit a PR.

**Comment by mleanos** [comment]
This looks like a good change. However, with some changes. I'd like to see a PR for this.

**Comment by mleanos** [comment]
A couple things that jump out at me..

1) The constructor should be returning the service, and not just the promise
2) I'd like to see `signin` & `signout` implemented with their own function for custom logic & handling of the resolved promise from the api call.
3) The `refresh` method should handle the resolving of the api call, so we can catch errors there & we don't need to set the services `user` field unless we get a response.
4) Add a `init()` method of the service

Have a look at this article. I found it quite interesting & I think we should be preparing ourselves for ES6. In particular, the points about Angular... http://blog.thoughtram.io/angularjs/es6/2015/01/23/exploring-angular-1.3-using-es6.html

I know you probably weren't looking for a review of that snippet. However, I just wanted to clarify my previous comment, before you did a bunch of work for a PR.

**Comment by trainerbill** [comment]
Angular Service?  I thought we were always using factories?

1)  It returns the service object when the promise resolves....  You have to return the promise first.
2)  Fair enough, but I would return the promise.
3)  I would agree that there needs to be a catch() initially and a reject if refresh fails.  Also only set service.user in then and have a catch there as well.
4)  init would do the same thing as refresh.  Rename?
5)  I don't know enough about ES6 to comment.  Think we should open a disscussion on it. 

So how this is going to work is here:

https://github.com/meanjs/mean/blob/master/modules/core/client/app/init.js#L19

We need to change to Authentication.then().catch() and then do the state checks.  So we need to return the promise right off the bat.

**Comment by mleanos** [comment]
@trainerbill Angular services are more suited for implementing ES6 classes. This is described a bit in the article I shared in my last comment.

I don't think the Authentication service should ever return a promise.. It should merely be a service with methods/members that may, or may not, return promises. I can't ever see an instance where an actual class (or service) constructor should return anything other than itself. 

If you need to the service to return a promise at any point, just wrap that in a property or method of the service.

Let's pick this discussion up a bit more once the Articles style guide PR is merged. We'll then have better context of how everything is going to be working with the new styles/implementation.

**Comment by trainerbill** [comment]
@mleanos Does using ES6 on the client side present browser compatibility issues?

**Comment by ilanbiala** [comment]
@trainerbill I think we should focus on other things rather than ES6 support for 0.5.0. Maybe for 0.6.0 it's worth looking into?

**Comment by trainerbill** [comment]
@ilanbiala I would agree for the client side, because I think we will reduce browser compatibility significantly.  But I think the Authentication structure as it currently is implemented needs to change.  Which was the purpose for this Issue.  I think we need to redo the AuthService using ES5 and worry about ES6 later.

**Comment by mleanos** [comment]
Don't get me wrong. I'm not advocating to implement ES6 features right now. I was merely making a point that Angular Services are better suited for the migration, once we make it. 

Since there isn't really a difference between Factories & Services, we should use the latter. There are benefits (other than ES6 related) for using Services. The article points some of the benefits out.

**Comment by trainerbill** [comment]
@mleanos fair enough.  So the point of the issue is discussing the Authentication service.  I would like some direction on where to go, because the current method is not ideal at all.

1.  Ignore all ES6 stuff :)
2.  Don't return a promise.  I am ok with not returning a promise.  easy fix would be to do user: refresh() to invoke the refresh function on instatiation.  Then you would have to do a Authentication.user.then() when waiting for authentication to resolve.  Not a huge deal
3.  Anything else?

**Comment by mleanos** [comment]
@trainerbill That sounds like a good plan. However, we should add the `me()` method to the existing Users service. I think that's where it belongs.

Couple of questions..
1. What is the intended purpose of the `refresh()` method? How does it differ from `me()`?
2. In general, what do we want to do with the Authentication service?
&nbsp; &nbsp; a. Signout, Signin are obvious methods
&nbsp; &nbsp; b. Should we have a method to perform the authentication, and separate that logic from the Signin logic?
3. One huge consideration is what the alternative to using `$window.user` is.
&nbsp; &nbsp; a. Do we look into client-side storage (ngStorage/Token/HTML5)? This might go hand-in-hand with whatever implementation we go with for Token-based Authentication.

I completely agree with you on the need for refactoring the Authentication service. It's something that has been bothering me for quite some time. I figured we wouldn't be able to avoid it once we moved to Token-based Authentication. 

Since this issue is very intimately related to Token-based Authentication, perhaps you want to get involved in that conversation; and maybe even you'd want to work on the client-side aspects of it. I've already started a PR to track it, and get feedback; I only touched the server-side of the implementation. I'd love to get that ball rolling again. https://github.com/meanjs/mean/pull/1121
